### PR TITLE
Add init container for mysql initialization

### LINKED ISSUE #229: Failed to initialize root password and custom user/password.
I created a mysql cluster, but the generated mysql cluster didn't initialize root with specifyed password and custom user/password were neither initialized.

mysql-operator: 0.2.2
```
mysql> select user, host from mysql.user;
+--------------+-----------+
| user         | host      |
+--------------+-----------+
| orchestrator | %         |
| repl_McYOj   | %         |
| exp_dnuFW    | 127.0.0.1 |
| root         | localhost |
+--------------+-----------+
```

cluster.yaml
```yaml
apiVersion: mysql.presslabs.org/v1alpha1
kind: MysqlCluster
metadata:
  name: mc
spec:
  # image: percona:5.7.24
  replicas: 2
  secretName: my-secret

  mysqlConf:
    slow_query_log: "1"
    log_output: table
    long_query_time: "20"
    slow_query_log_file: /dev/stdout

  podSpec:
     resources:
       requests:
         memory: 256Mi
         cpu:    200m

  ## Specify additional volume specification
  volumeSpec:
     accessModes: [ "ReadWriteOnce" ]
     resources:
       requests:
         storage: 1Gi
```

secret.yaml
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: my-secret
  namespace: default
type: Opaque
data:
  # root password is required to be specified
  ROOT_PASSWORD: cm9vdA==
  USER: Ymx1dmVu
  PASSWORD: eWFuc2hpeWk=
```

Partial of mysql node env
```
MYSQL_PASSWORD=yanshiyi
MYSQL_OPERATOR_TEST_ORCHESTRATOR_PORT=tcp://10.102.244.101:80
MYSQL_ROOT_PASSWORD=root
MYSQL_USER=bluven
```

### PR DESCRIPTION
**Pull request by AMecea**
Fixes #229

### CODE REVIEW

File: pkg/controller/mysqlcluster/internal/syncer/statefullset.go [hunk 1]
```diff
@@ -569,8 +597,24 @@ func (s *sfsSyncer) getVolumeMountsFor(name string) []core.VolumeMount {
...
+			core.VolumeMount{
+				Name:      confVolumeName,
+				MountPath: ConfVolumeMountPath,
+			},
+			core.VolumeMount{
+				Name:      dataVolumeName,
+				MountPath: DataVolumeMountPath,
+			},
+			core.VolumeMount{
+				Name:      onlyInitCustomScriptVolumeName,
```
Comment (calind) [inline pkg/controller/mysqlcluster/internal/syncer/statefullset.go]:
```suggestion
				Name:      initVolumeName,
```

File: pkg/controller/mysqlcluster/internal/syncer/statefullset.go [hunk 2]
```diff
@@ -569,8 +597,24 @@ func (s *sfsSyncer) getVolumeMountsFor(name string) []core.VolumeMount {
...
+				MountPath: onlyInitCustomScriptPath,
```
Comment (calind) [inline pkg/controller/mysqlcluster/internal/syncer/statefullset.go]:
```suggestion
				MountPath: initVolumeMountPath,
```

File: pkg/controller/mysqlcluster/internal/syncer/settings.go [hunk 3]
```diff
@@ -65,4 +65,10 @@ var (
 
 	// ExporterTargetPort is the port for the exporter
 	ExporterTargetPort = intstr.FromInt(ExporterPort)
+
+	onlyInitScriptName = "onlyInit.sh"
```
Comment (calind) [inline pkg/controller/mysqlcluster/internal/syncer/settings.go]:
```suggestion
var initScripts = map[string]string{
"99-exit-when-done.sh": `
echo "Initialization complete, now exiting!"
exit 0
`)
}
```

File: pkg/controller/mysqlcluster/internal/syncer/settings.go [hunk 4]
```diff
@@ -65,4 +65,11 @@ var (
 
 	// ExporterTargetPort is the port for the exporter
 	ExporterTargetPort = intstr.FromInt(ExporterPort)
+
+	initScripts = map[string]string{
+		"99-exit-when-donw.sh": `
```
Comment (calind) [inline pkg/controller/mysqlcluster/internal/syncer/settings.go]:
```suggestion
		"99-exit-when-done.sh": `
```

File: pkg/controller/mysqlcluster/internal/syncer/statefullset.go [hunk 5]
```diff
@@ -41,12 +41,16 @@ const (
...
+	initVolumeName = "custom-init"
+	initScriptPath = "/docker-entrypoint-initdb.d"
 )
 
 // containers names
 const (
-	containerInitName      = "init-mysql"
+	containerInitName      = "init-conf"
 	containerCloneName     = "clone-mysql"
+	containerMysqlInitName = "init-mysql"
```
Comment (calind) [inline pkg/controller/mysqlcluster/internal/syncer/statefullset.go]:
```suggestion
	containerMysqlBootstrapName = "bootstrap-mysql"
```

File: pkg/controller/mysqlcluster/internal/syncer/statefullset.go [hunk 6]
```diff
@@ -473,6 +494,17 @@ func (s *sfsSyncer) ensureVolumes() []core.Volume {
...
 		ensureVolume(dataVolumeName, dataVolume),
+
+		ensureVolume(initVolumeName, core.VolumeSource{
+			ConfigMap: &core.ConfigMapVolumeSource{
+				LocalObjectReference: core.LocalObjectReference{
+					Name: s.cluster.GetNameForResource(mysqlcluster.ConfigMap),
+				},
+				Items:       initVolumeItems,
+				DefaultMode: &execFileMode,
+				Optional:    &trueVar,
```
Comment (calind) [inline pkg/controller/mysqlcluster/internal/syncer/statefullset.go]: Why is this optional?

**Comment by AMecea** [comment]
This approach does not work. 

When the pod gets restarted the init container will skip the initialization and the custom script `exist when done` will not be run so the mysqld is started and pods get stuck in init phase.

**Pull request closed without merging**